### PR TITLE
Remove proptypes in production

### PR DESCRIPTION
**Pull request by mbrevda**
As react doesn't check proptypes in production, there is no point in keeping them in the build. Removing them make the build size ever so slightly smaller.

~~While I didn't see tests passing locally, they don't seem to be passing on master, either.~~ If CI says its good, then its good!

### CODE REVIEW

**Comment by bvaughn** [comment]
Hey @mbrevda!

Stripping `propTypes` in production is a great idea- but I think it belongs in application land rather than in react-virtualized. (If `propTypes` are stripped from RV, then users won't get helpful validation messages when developing their applications using RV.)

Fortunately this can be accomplished after the fact (and easier!) using the [babel-plugin-transform-react-remove-prop-types](https://github.com/oliviertassinari/babel-plugin-transform-react-remove-prop-types) plug-in.

Out of curiosity, I ran it over `Grid` and the CommonJS version shrinks from 40k to 35k. So it's definitely worth doing (as an application build step)!

**Comment by mbrevda** [comment]
If I may make a (possible erroneous) counter argument: running `npm run
build` didn't seem to strip the proptypes from the built dist files. This
means that devs would still be able to take advantage of them AND have them
auto stripped if they follow typical react good practices - all without
having to add extra globally affecting plugins.

**Comment by bvaughn** [comment]
> running `npm run build` didn't seem to strip the proptypes from the built dist files.

What do you mean? Within react-virtualized? That's because I don't use that plugin to strip them (because I think they're useful to leave in place for application developers). But they can _later_ be stripped when the application is built using a plug-in like the one I mentioned above.

I'm afraid I don't understand the counter-argument?

**Comment by mbrevda** [comment]
> That's because I don't use that plugin to strip them

Good, then they will be there during development.

> when the application is built using a plug-in

That's assuming a plugin is, indeed, used. It's also assuming that babel
will operate on `node_modules` (which it usually doesn't).

Wouldn't it be simpler - and just as useful - to surround proptypes with a
NODE_ENV check?

**Comment by bvaughn** [comment]
>>  when the application is built using a plug-in

> That's assuming a plugin is, indeed, used. It's also assuming that babel
will operate on `node_modules` (which it usually doesn't).

Ah! I understand the point you're making now. Okay, let me re-open this PR and think about it. I'm not a fan of moving the props outside of the class (just a subjective thing).

**Comment by bvaughn** [comment]
This discussion thread seems relevant: https://github.com/oliviertassinari/babel-plugin-transform-react-remove-prop-types/issues/38

**Comment by bvaughn** [comment]
I reached out to Olivier (author of the Babel plugin I mentioned above) asking for his recommendations. (He's also a contributor to Material UI so I figured he's thought about this from the library angle a lot already.)

He's going to get back to me with a more detailed answer tomorrow, but in passing he said:
> I believe that the wrap mode is used in the v0.16.0 version of Material-UI right now in production. Also I had to fix one issue, I'm not sure I have released it.

So I'd probably like to wait until I hear his thoughts too and then we can decide the best way to move forward. 😄

**Comment by bvaughn** [comment]
Reached out for some input on Twitter ([here](https://twitter.com/brian_d_vaughn/status/787033766774476801)) and it looks like your suggested approach is right on the money.

**Comment by bvaughn** [comment]
Did a practical experiment to see how much space would be saved by a typical react-virtualized usage.

```js
// Using create-react-app and the following App.js code
class App extends Component {
  render() {
    return (
      <Grid />
    );
  }
}

// Importing Grid this way ~> 252 kb
import { Grid } from 'react-virtualized';

// Importing Grid this way ~> 186 kb
import Grid from 'react-virtualized/dist/commonjs/Grid/Grid';

// propTypes wrapped with NODE_ENV ~> 185 kb
import Grid from 'react-virtualized/dist/commonjs/Grid/Grid';
```

Doing the same as above but using `AutoSizer` + `List` + `Grid` (a more typical sample) the sizes are: 252 kb, 197kb, and 195 kb.

So the size savings aren't huge but they're definitely significant enough to warrant the change.

File: source/Table/Column.js [hunk 1]
```diff
@@ -77,3 +67,14 @@ export default class Column extends Component {
     width: PropTypes.number.isRequired
   }
 }
+
+if (process.env.NODE_ENV !== 'production') {
```
Comment (bvaughn) [inline source/Table/Column.js]: This should be wrapping `propTypes` not `defaultProps`. Those should stay in the class.

**Comment by bvaughn** [comment]
Looks like `Table` and `SortIndicator` got skipped and `Column` got shuffled slightly (`defaultProps` got wrapped instead of `propTypes`).

**Comment by bvaughn** [comment]
I'll go ahead and address those minor issues

**Comment by bvaughn** [comment]
This change just went out as 8.1.1. Gave you thanks in the CHANGELOG. Cheers!

**Comment by mbrevda** [comment]
Belated response:
> let me re-open this PR

Thanks!

> So the size savings aren't huge

Wasn't expecting much, but if every package reduced just a few bytes...

> This should be wrapping propTypes not defaultProps

Oops, sorry

> I'll go ahead and address those minor issues

Thanks!

> This change just went out

First PR! 🎊🎉 Awesome!

**Pull request merged**